chainer_chemistry/models/gat.py

```diff
@@ -0,0 +1,169 @@
+        hidden_dim (int): dimension of feature vector
+            associated to each atom
+        n_layers (int): number of layers
+        n_atom_types (int): number of types of atoms
+        concat_hidden (bool): If set to True, readout is executed in each layer
+            and the result is concatenated
+        weight_tying (bool): enable weight_tying or not
+
+    """
+    # NUM_EDGE_TYPE = 4
+
+    def __init__(self, out_dim, hidden_dim=16, heads=2, negative_slope=0.2,
+                 n_layers=4, n_atom_types=MAX_ATOMIC_NUM, concat_hidden=False,
+                 weight_tying=True):
+        super(GraphAttentionNetworks, self).__init__()
+        n_readout_layer = n_layers if concat_hidden else 1
+        # n_message_layer = 1 if weight_tying else n_layers
+        n_message_layer = n_layers
+        with self.init_scope():
+            # Update
+            self.embed = EmbedAtomID(out_size=hidden_dim, in_size=n_atom_types)
+            # self.weight = GraphLinear(hidden_dim, heads * hidden_dim)
+            # self.att_weight = GraphLinear(hidden_dim * 2, 1)
+            self.message_layers = chainer.ChainList(
+                *[GraphLinear(hidden_dim, heads * hidden_dim)
+                  for _ in range(n_message_layer)]
+            )
+            self.attenstion_layers = chainer.ChainList(
+                *[GraphLinear(hidden_dim * 2, 1)
+                  for _ in range(n_message_layer)]
+            )
+            # self.update_layer = links.GRU(2 * hidden_dim, hidden_dim)
+            # Readout
+            self.i_layers = chainer.ChainList(
+                *[GraphLinear(2 * hidden_dim, out_dim)
+                    for _ in range(n_readout_layer)]
+            )
+            self.j_layers = chainer.ChainList(
+                *[GraphLinear(hidden_dim, out_dim)
+                    for _ in range(n_readout_layer)]
+            )
+        self.out_dim = out_dim
+        self.heads = heads
+        self.hidden_dim = hidden_dim
+        self.n_layers = n_layers
+        self.concat_hidden = concat_hidden
+        self.weight_tying = weight_tying
+        self.negative_slope = negative_slope
+
+    def update(self, h, adj, step=0):
+        xp = self.xp
+        # (minibatch, atom, channel)
+        mb, atom, ch = h.shape
+        # (minibatch, atom, heads * out_dim)
+        h = self.message_layers[step](h)
+        # (minibatch, atom, heads, out_dim)
+        h = functions.reshape(h, (mb, atom, self.heads, self.hidden_dim))
+
+        # concat all pairs of atom
+        # (minibatch, 1, atom, heads, out_dim)
+        h_i = functions.expand_dims(h, axis=1)
+        # (minibatch, atom, atom, heads, out_dim)
+        h_i = functions.broadcast_to(h_i, (mb, atom, atom,
+                                           self.heads, self.hidden_dim))
+        device_id = cuda.get_device_from_array(h_i.array).id
+        h_j = functions.copy(h_i, device_id)
+        h_j = functions.transpose(h_j, (0, 2, 1, 3, 4))
```

Comment:
this is completely matter of preference, but I feel it is easy to understand to create `h_j` same way with `h_i` using `expand_dims` and `broadcast_to`.

Reply:
Sorry. I would like to make sure.
Are these codes enough?
`h_j = functions.expand_dims(h, axis=2)` 
`h_j = functions.broadcast_to(h_i, (mb, atom, atom, self.heads, self.hidden_dim))`